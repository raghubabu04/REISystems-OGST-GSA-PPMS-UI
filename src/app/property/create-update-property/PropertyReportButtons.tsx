import React from "react";
import { PPMSButton } from "../../../ui-kit/components/common/PPMS-button";
import { PPMSSubmit } from "../../../ui-kit/components/PPMS-submit";

interface PropertyReportButtonProps {
  saveFunction: () => void;
  isSubmitDisabled: boolean;
  isSubmitLoading: boolean;
  cancelFunction: () => void;
  workflow?: any;
  saveId?: string;
  cancelId?: string;
  approveId?: string;
  isSaveAndReportDisplayed: boolean;
  saveAndReportAgainId?: string;
  saveAndReportAgain?: any;
}

export default function PropertyReportButtons(
  props: PropertyReportButtonProps
) {
  return (
    <>
      <div className={"margin-right-auto col-md-auto"}>
        <ul className="usa-button-group usa-button-group--segmented">
          <li className="usa-button-group__item ">
            <PPMSButton
              variant={"primary"}
              type={"button"}
              value={""}
              label={"Save"}
              onPress={props.saveFunction}
              id={props.saveId}
              className={"out-button"}
            />
          </li>
          {props.workflow === "APPROVE_FOREIGN_GIFT" ? (
            <li className="usa-button-group__item">
              <PPMSButton
                variant={"primary"}
                type={"submit"}
                value={"submit"}
                label={"Approve"}
                isDisabled={props.isSubmitDisabled}
                isLoading={props.isSubmitLoading}
                id={props.approveId}
              />
            </li>
          ) : (
            <li className="usa-button-group__item">
              <PPMSButton
                variant={"primary"}
                type={"submit"}
                label={"Submit"}
                className={"out-button"}
                isDisabled={props.isSubmitDisabled}
                isLoading={props.isSubmitLoading}
                id={"property-submit"}
              />
            </li>
          )}
          <li className="usa-button-group__item">
            <PPMSButton
              variant={"primary"}
              type={"reset"}
              label={"Cancel"}
              onPress={props.cancelFunction}
              id={props.cancelId}
              className={"out-button"}
            />
          </li>
        </ul>
      </div>
      {props.isSaveAndReportDisplayed ? (
      <div className="grid-col align-self-right submit-report-padding">
        <PPMSButton
          variant={"primary"}
          type={"submit"}
          label={"Submit/Report Similar"}
          onPress={props.saveAndReportAgain}
          id={props.saveAndReportAgainId}
          className={"out-button"}
        />
      </div>
      ) : ""}
    </>
  );
}
