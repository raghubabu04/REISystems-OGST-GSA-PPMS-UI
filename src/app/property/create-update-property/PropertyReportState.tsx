import {FSCModel} from "./federal-supply-class/model/FSCModel";

export interface PropertyReportState {
  showErrorAlert?: boolean;
  FormErrorMessage?: string;
  alertBodyArray?: string[];
  isSubmitDisabled?: boolean;
  isSubmitLoading?: boolean;
  isFormValidated?: boolean;
  verifyDisabled?: boolean;
  fsc?: FSCModel;
  isSubmitted?: boolean;
  isDraft?: boolean;
  propertyId?: number;
  stateValues?: any[];
  triggerValidation?: boolean;
  fileInfectedStatus?: boolean;
  submittedDate?: Date;
  accordion: {
    toggleAllAccordion: boolean;
    toggleICNAccordion: boolean;
    toggleAgencyInfoAccordion: boolean;
    toggleReportingAgencyAddressAccordion: boolean;
    togglePointofContactAccordion: boolean;
    togglePropertyLocationAccordion: boolean;
    togglePropertyCustodianAccordion: boolean;
    togglePropertyInformationAccordion: boolean;
    toggleFederalSupplyClassAccordion: boolean;
    toggleQuantityAndCostAccordion: boolean;
    toggleAdditionalInformationAccordion: boolean;
    toggleSalesNotesAccordion: boolean;
    toggleUploadMultiplPicturesDocumentsAccordion: boolean;
    toggleGiftInfoAccordion: boolean;
    toggleDonorInfoAccordian: boolean;
    toggleRecipientInfoAccordian: boolean;
    toggleAppraisalInfoAccordion: boolean;
    toggleAppraisalAgencyInformationAccordian: boolean;
    openItems: any[];
  };
  propertyGroup?: string;
  showActionHistoryModal?: boolean;
  showAddToPropertyActionModal: boolean;
  actionHistoryData: any;
  propertyData: any;
  showReportAgain: boolean;
  reportAgain:boolean;
}

export const PropertyReportStateDefault = {
  showErrorAlert: false,
  FormErrorMessage: "",
  alertBodyArray: [],
  isSubmitLoading: false,
  isSubmitDisabled: false,
  isFormValidated: false,
  verifyDisabled: false,
  fsc: new FSCModel(),
  isSubmitted: false,
  isDraft: true,
  propertyId: null,
  stateValues: [],
  triggerValidation: false,
  fileInfectedStatus: false,
  submittedDate: null,
  accordion: {
    toggleAllAccordion: true,
    toggleICNAccordion: true,
    toggleAgencyInfoAccordion: true,
    toggleReportingAgencyAddressAccordion: true,
    togglePointofContactAccordion: true,
    togglePropertyLocationAccordion: true,
    togglePropertyCustodianAccordion: true,
    togglePropertyInformationAccordion: true,
    toggleFederalSupplyClassAccordion: true,
    toggleQuantityAndCostAccordion: true,
    toggleAdditionalInformationAccordion: true,
    toggleSalesNotesAccordion: true,
    toggleUploadMultiplPicturesDocumentsAccordion: true,
    toggleGiftInfoAccordion: true,
    toggleDonorInfoAccordian: true,
    toggleRecipientInfoAccordian: true,
    toggleAppraisalInfoAccordion: true,
    toggleAppraisalAgencyInformationAccordian: true,
    openItems: [
      "toggleICNAccordion",
      "toggleAgencyInfoAccordion",
      "toggleReportingAgencyAddressAccordion",
      "togglePointofContactAccordion",
      "togglePropertyLocationAccordion",
      "togglePropertyCustodianAccordion",
      "togglePropertyInformationAccordion",
      "toggleFederalSupplyClassAccordion",
      "toggleQuantityAndCostAccordion",
      "toggleAdditionalInformationAccordion",
      "toggleSalesNotesAccordion",
      "toggleUploadMultiplPicturesDocumentsAccordion",
      "toggleGiftInfoAccordion",
      "toggleDonorInfoAccordian",
      "toggleRecipientInfoAccordian",
      "toggleAppraisalAgencyInformationAccordian",
      "toggleAppraisalInfoAccordion",
      "toggleAppraisalAgencyInformationAccordian",
    ],
  },
  propertyGroup: "",
  showAddToPropertyActionModal: false,
  actionHistoryData: [],
  showActionHistoryModal: false,
  propertyData: "",
  showReportAgain: false,
  reportAgain: false,
};
