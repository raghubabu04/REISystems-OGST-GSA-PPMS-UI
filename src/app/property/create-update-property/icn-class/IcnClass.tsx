import React, { useContext, useMemo } from "react";
import { PropertyContext } from "../PropertyContext";
import { IcnState } from "./IcnState";
import AacCode from "./AacCode";
import { validateAACRistrictionInSalesCenter } from "../validations/validateFSCSalesCenter";
import { validatePropertyField } from "../../../../ui-kit/components/validations/FieldValidations";
import { UserUtils } from "../../../../utils/UserUtils";
import { PropertyApiService } from "../../../../api-kit/property/property-api-service";
import { PPMSTooltip } from "../../../../ui-kit/components/common/PPMS-tooltip";
import { PPMSInput } from "../../../../ui-kit/components/common/input/PPMS-input";
import { PPMSAlert } from "../../../../ui-kit/components/common/PPMS-alert";
import { history } from "../../../../_redux/_helpers/history";
import { PropertyGroupType } from "../constants/Constants";

interface IcnProps {
  resetOptions?: any;
  propertyGroup?: any;
}

function IcnClass(props: IcnProps) {
  let propertyApiService = new PropertyApiService();
  const {
    icnState,
    updateIcnState,
    propertyInfoState,
    agencyInfoState,
    updateFSCState,
    updatePropertyReportState,
  } = useContext(PropertyContext);

  function handleSerialNumChange(event: any) {
    const validation = validatePropertyField(
      event.target.id,
      icnState.serialNum
    );
    updateIcnState({
      // serialNum: event.target.value.toUpperCase(),
      serialValidationMessage: validation.validationError,
      serialIsInvalid: validation.isInvalid,
      serialIsValid: !validation.isInvalid,
      showIcnCodeValid: false,
    });
  }

  function handleSuffixChange(event: any) {
    const validation = validatePropertyField(event.target.id, icnState.suffix);
    updateIcnState({
      suffix: event.target.value.toUpperCase(),
      suffixValidationMessage: validation.validationError,
      suffixIsInvalid: validation.isInvalid,
      suffixIsValid: !validation.isInvalid,
      showIcnCodeValid: false,
    });
  }

  function checkIfValuesAreEntered() {
    if (icnState.aacCode === "") {
      updateIcnState({
        aacIsInvalid: true,
        aacValidationMessage:
          "Please enter all 6 characters of Activity Address Code.",
      });
    }
    if (icnState.julianDate === "") {
      updateIcnState({
        julianDateIsInvalid: true,
        julianDateValidationMessage: "Julian date must be 4 numbers",
      });
    }
    if (icnState.serialNum === "") {
      updateIcnState({
        serialValidationMessage: "Serial Number must be 4 Characters",
        serialIsInvalid: true,
      });
    }
  }

  function handleConfirmICN() {
    props.resetOptions();
    if (!icnState.isIcnValid) {
      if (
        icnState.aacIsInvalid ||
        icnState.aacCode === "" ||
        icnState.julianDateIsInvalid ||
        icnState.julianDate === "" ||
        icnState.serialIsInvalid ||
        icnState.serialNum === "" ||
        icnState.suffixIsInvalid
      ) {
        checkIfValuesAreEntered();
        updateIcnState({
          showForm: false,
        });
      } else {
        updateIcnState({
          disableIcnConfirmButton: true,
        });

        let savePropertyReportBody = {
          itemControlNumber: joinIcn(icnState),
          aacId: icnState.aacCode,
          agencyBureau: agencyInfoState.aac,
        };

        if (props.propertyGroup === PropertyGroupType.FOREIGN_GIFT) {
          savePropertyReportBody = {
            ...savePropertyReportBody,
            ...{
              propertyGroup: props.propertyGroup,
              recipientInfo: {
                email: "",
                firstName: "",
                lastName: "",
                phone: "",
                phoneExtension: "",
                recipientTitle: "",
              },
              giftInfo: {
                administration: "",
                dateReceivedByFgm: null,
                dateReceivedByRecipient: null,
                fairMarketValue: null,
                fiscalYear: "",
                giftFscCode: "",
                recipientWantsToBuy: null,
                unitOfIssue: "EA",
                upsetPrice: null,
                vaultLocation: "",
                vaultShelfNumber: "",
              },
            },
          };
        } else {
          savePropertyReportBody = {
            ...savePropertyReportBody,
            ...{ propertyGroup: "property" },
          };
        }

        propertyApiService
          .verifyICN(joinIcn(icnState))
          .then((response) => {
            if (response.data.status === "ICN NOT FOUND") {
              propertyApiService
                .savePropertyReport(savePropertyReportBody)
                .then((response) => {
                  updatePropertyReportState({
                    propertyId: response.data.propertyId,
                  });
                  updateIcnState({
                    disableAAC: true,
                    disableJulianDate: true,
                    disableSerialNumber: true,
                    disableSuffix: true,
                    showIcnCodeValid: true,
                    isIcnValid: true,
                    icnVerificationMessage: "ICN is Confirmed",
                  });
                })
                .catch(() => {
                  updateIcnState({
                    showIcnCodeValid: true,
                    isIcnValid: false,
                    icnVerificationMessage:
                      "Error Confirming ICN. Please try again.",
                  });
                });
            } else if (response.data.status === "ICN Found") {
              updateIcnState({
                showIcnCodeValid: true,
                isIcnValid: false,
                icnVerificationMessage:
                  "ICN already exists in the System. Please enter another ICN.",
              });
            }
          })
          .catch(() => {
            updateIcnState({
              showIcnCodeValid: true,
              isIcnValid: false,
              icnVerificationMessage: "Error Confirming ICN. Please try again.",
            });
          })
          .finally(() => {
            updateIcnState({
              disableIcnConfirmButton: false,
            });
          });
      }
    }

    // PPDMS:763 Validate against ICN AAC=47310K
    let isAACCodeRestrictedInSalesCenter = validateAACRistrictionInSalesCenter(
      icnState.aacCode
    );

    if (isAACCodeRestrictedInSalesCenter) {
      updateFSCState({
        federalSalesCenter: "G",
        isSalesCenterdisabled: true,
      });
    }
  }

  function joinIcn(state: IcnState) {
    let icnVal = "";
    if (icnState.suffix) {
      icnVal = `${icnState.aacCode}${icnState.julianDate}${icnState.serialNum}${icnState.suffix}`;
    } else {
      icnVal = `${icnState.aacCode}${icnState.julianDate}${icnState.serialNum}`;
    }
    return icnVal;
  }

  function handleJulianDateChange(event: any) {
    const validation = validatePropertyField(
      event.target.id,
      event.target.value
    );
    updateIcnState({
      julianDate: event.target.value,
      julianDateValidationMessage: validation.validationError,
      julianDateIsInvalid: validation.isInvalid,
      julianDateIsValid: !validation.isInvalid,
      showIcnCodeValid: false,
    });
  }

  return useMemo(() => {
    return (
      <React.Fragment>
        <div className={"grid-row grid-gap-4"}>
          <div className={"tablet:grid-col-2"}>
            <AacCode />
          </div>
          <div className={"tablet:grid-col-2"}>
            <PPMSInput
              id={"julian-date"}
              name={"julianDate"}
              label={"Julian date"}
              maxLength={4}
              onChange={handleJulianDateChange}
              validationMessage={icnState.julianDateValidationMessage}
              isRequired={true}
              isDisabled={
                !icnState.disableJulianDate
                  ? !(
                      UserUtils.hasPermission("APO") ||
                      UserUtils.hasPermission("NU") ||
                      UserUtils.hasPermission("SM")
                    )
                  : true
              }
              value={icnState.julianDate}
              inputType={"text"}
              isInvalid={icnState.julianDateIsInvalid}
              isValid={
                history?.location?.pathname === "/createPropertyReport"
                  ? icnState.julianDateIsValid
                  : !icnState.disableJulianDate
              }
            />
          </div>
          <div className={"tablet:grid-col-2"}>
            <PPMSInput
              id={"serial-number"}
              name={"serialNum"}
              label={"Serial Number"}
              maxLength={4}
              onChange={(event) => {
                updateIcnState({
                  serialNum: event.target.value.toUpperCase(),
                });
              }}
              onBlur={handleSerialNumChange}
              isRequired={true}
              validationMessage={icnState.serialValidationMessage}
              isInvalid={icnState.serialIsInvalid}
              isValid={icnState.serialIsValid}
              isDisabled={icnState.disableSerialNumber}
              inputType={"text"}
              value={icnState.serialNum}
            />
          </div>
          <div className={"tablet:grid-col-2"}>
            <PPMSInput
              id={"suffix"}
              name={"suffix"}
              label={"Suffix"}
              maxLength={1}
              onChange={(event) => {
                updateIcnState({
                  suffix: event.target.value.toUpperCase(),
                });
              }}
              onBlur={handleSuffixChange}
              validationMessage={icnState.suffixValidationMessage}
              isInvalid={icnState.suffixIsInvalid}
              isValid={icnState.suffixIsValid}
              isDisabled={icnState.disableSuffix}
              inputType={"text"}
              isRequired={false}
              value={icnState.suffix}
            />
          </div>
        </div>
        <div className={"grid-row update-buttons"}>
          {!icnState.isIcnValid && (
            <div className={"tablet:grid-col-4"}>
              <PPMSTooltip
                trigger={"focus"}
                id={"confirm-icn"}
                placement={"right"}
                tooltipContent={"ICN cannot be changed once confirmed."}
                triggerSource={
                  <button
                    className="usa-button"
                    type="button"
                    disabled={icnState.disableIcnConfirmButton}
                    onClick={handleConfirmICN}
                  >
                    Confirm ICN
                  </button>
                }
              />
            </div>
          )}
          <div className={"tablet:grid-col-8"}>
            <PPMSAlert
              id={"icn-verification-danger-msg"}
              show={icnState.showIcnCodeValid && !icnState.isIcnValid}
              alertBody={icnState.icnVerificationMessage}
              alertClassName={"email-verification-error"}
              alertKey={"email-verification-error"}
              alertVariant={"danger"}
            />
            <PPMSAlert
              id={"icn-verification-success-msg"}
              show={icnState.showIcnCodeValid && icnState.isIcnValid}
              alertBody={icnState.icnVerificationMessage}
              alertClassName={"email-verification-error"}
              alertVariant={"success"}
              alertKey={"email-verification-error"}
            />
          </div>
        </div>
      </React.Fragment>
    );
  }, [icnState, propertyInfoState]);
}

export default IcnClass;
