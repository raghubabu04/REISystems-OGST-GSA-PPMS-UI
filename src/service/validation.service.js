import { BehaviorSubject } from "rxjs";
const isFormSubmitted = new BehaviorSubject(false);
const validationService = {
  send: function (msg) {
    isFormSubmitted.next(msg);
  },
};

export { validationService, isFormSubmitted };
