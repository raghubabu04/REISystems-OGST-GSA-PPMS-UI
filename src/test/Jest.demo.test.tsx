import { mount, ReactWrapper, HTMLAttributes } from "enzyme";

import React from "react";
import { CommonApiService } from "../api-kit/common/common-api.service";
import IcnClass from "../app/property/create-update-property/icn-class/IcnClass";
import { PropertyProvider } from "../app/property/create-update-property/PropertyContext";

const icnClassComponent: ReactWrapper = mount(
  <PropertyProvider>
    <IcnClass />
  </PropertyProvider>
);

const mock: jest.SpyInstance = jest.spyOn(
  CommonApiService.prototype,
  "getBureau"
);

describe("component: item-control-number", () => {
  //when test, we need to create the update snapshot so the Julian Date matches. Otherwise, it will fail as the test date past the snapshot date.
  it("rendered", () => {
    expect(icnClassComponent).toMatchSnapshot();
  });
  describe(`field: aac-code`, () => {
    const field: ReactWrapper<HTMLAttributes, any> = icnClassComponent.find(
      `input[name="aacCode"]`
    );

    it("rendered field", () => {
      expect(field.length).toEqual(1);
    });

    const node = field.get(0);
    it(`has maxLength 6`, () => {
      expect(node.props.maxLength).toBe(6);
    });
    it("is required", () => {
      expect(node.props.required).toBe(true);
    });
    it(`Basic validation for AAC-Code.`, () => {
      field.simulate("change", { target: { value: "123#678" } });
      const fieldError = icnClassComponent.find(`#errorMessage-aac-code`);
      const node = fieldError.get(0);
      expect(node.props.children).toBe("Special characters are not allowed.");
    });
  });
});
