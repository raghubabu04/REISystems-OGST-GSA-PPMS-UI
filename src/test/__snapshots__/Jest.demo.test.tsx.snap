// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: item-control-number rendered 1`] = `
<PropertyProvider>
  <IcnClass>
    <div
      className="grid-row grid-gap-1"
    >
      <div
        className="tablet:grid-col-3"
      >
        <AacCode>
          <PPMSInput
            id="aac-code"
            inputType="text"
            isDisabled={false}
            isInvalid={false}
            isRequired={true}
            isValid={false}
            label="AAC CODE"
            maxLength={6}
            name="aacCode"
            onBlur={[Function]}
            onChange={[Function]}
            validationMessage="Is Required"
            value=""
          >
            <PPMSFormGroup
              error={false}
              id="formGroup-aac-code"
            >
              <div
                className="usa-form-group"
                data-testid="formGroup"
                id="formGroup-aac-code"
              >
                <PPMSLabel
                  hint=""
                  htmlFor="aac-code"
                >
                  <label
                    className="usa-label"
                    data-testid="label"
                    htmlFor="aac-code"
                  >
                     
                    AAC CODE
                     
                     
                  </label>
                </PPMSLabel>
                <PPMSTextInput
                  autoComplete="off"
                  disabled={false}
                  id="aac-code"
                  inputRef={
                    Object {
                      "current": <input
                        autocomplete="off"
                        class="usa-input"
                        data-testid="textInput"
                        id="aac-code"
                        maxlength="6"
                        name="aacCode"
                        required=""
                        type="text"
                        value=""
                      />,
                    }
                  }
                  maxLength={6}
                  name="aacCode"
                  onBlur={[Function]}
                  onChange={[Function]}
                  required={true}
                  type="text"
                  validationStatus=""
                  value=""
                >
                  <input
                    autoComplete="off"
                    className="usa-input"
                    data-testid="textInput"
                    disabled={false}
                    id="aac-code"
                    key="key-aac-code"
                    maxLength={6}
                    name="aacCode"
                    onBlur={[Function]}
                    onChange={[Function]}
                    required={true}
                    type="text"
                    value=""
                  />
                </PPMSTextInput>
              </div>
            </PPMSFormGroup>
          </PPMSInput>
        </AacCode>
      </div>
      <div
        className="tablet:grid-col-3"
      >
        <PPMSInput
          id="julian-date"
          inputType="text"
          isDisabled={true}
          isInvalid={false}
          isRequired={true}
          isValid={true}
          label="Julian date"
          maxLength={4}
          name="julianDate"
          onChange={[Function]}
          validationMessage="Is Required"
          value="1028"
        >
          <PPMSFormGroup
            error={false}
            id="formGroup-julian-date"
          >
            <div
              className="usa-form-group"
              data-testid="formGroup"
              id="formGroup-julian-date"
            >
              <PPMSLabel
                hint=""
                htmlFor="julian-date"
              >
                <label
                  className="usa-label"
                  data-testid="label"
                  htmlFor="julian-date"
                >
                   
                  Julian date
                   
                   
                </label>
              </PPMSLabel>
              <PPMSTextInput
                autoComplete="off"
                disabled={true}
                id="julian-date"
                inputRef={
                  Object {
                    "current": <input
                      autocomplete="off"
                      class="usa-input"
                      data-testid="textInput"
                      disabled=""
                      id="julian-date"
                      maxlength="4"
                      name="julianDate"
                      required=""
                      type="text"
                      value="1028"
                    />,
                  }
                }
                maxLength={4}
                name="julianDate"
                onBlur={[Function]}
                onChange={[Function]}
                required={true}
                type="text"
                validationStatus=""
                value="1028"
              >
                <input
                  autoComplete="off"
                  className="usa-input"
                  data-testid="textInput"
                  disabled={true}
                  id="julian-date"
                  key="key-julian-date"
                  maxLength={4}
                  name="julianDate"
                  onBlur={[Function]}
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value="1028"
                />
              </PPMSTextInput>
            </div>
          </PPMSFormGroup>
        </PPMSInput>
      </div>
      <div
        className="tablet:grid-col-3"
      >
        <PPMSInput
          id="serial-number"
          inputType="text"
          isDisabled={false}
          isInvalid={false}
          isRequired={true}
          isValid={false}
          label="Serial Number"
          maxLength={4}
          name="serialNum"
          onBlur={[Function]}
          onChange={[Function]}
          validationMessage="Is Required"
          value=""
        >
          <PPMSFormGroup
            error={false}
            id="formGroup-serial-number"
          >
            <div
              className="usa-form-group"
              data-testid="formGroup"
              id="formGroup-serial-number"
            >
              <PPMSLabel
                hint=""
                htmlFor="serial-number"
              >
                <label
                  className="usa-label"
                  data-testid="label"
                  htmlFor="serial-number"
                >
                   
                  Serial Number
                   
                   
                </label>
              </PPMSLabel>
              <PPMSTextInput
                autoComplete="off"
                disabled={false}
                id="serial-number"
                inputRef={
                  Object {
                    "current": <input
                      autocomplete="off"
                      class="usa-input"
                      data-testid="textInput"
                      id="serial-number"
                      maxlength="4"
                      name="serialNum"
                      required=""
                      type="text"
                      value=""
                    />,
                  }
                }
                maxLength={4}
                name="serialNum"
                onBlur={[Function]}
                onChange={[Function]}
                required={true}
                type="text"
                validationStatus=""
                value=""
              >
                <input
                  autoComplete="off"
                  className="usa-input"
                  data-testid="textInput"
                  disabled={false}
                  id="serial-number"
                  key="key-serial-number"
                  maxLength={4}
                  name="serialNum"
                  onBlur={[Function]}
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value=""
                />
              </PPMSTextInput>
            </div>
          </PPMSFormGroup>
        </PPMSInput>
      </div>
      <div
        className="tablet:grid-col-3"
      >
        <PPMSInput
          id="suffix"
          inputType="text"
          isDisabled={false}
          isInvalid={false}
          isRequired={false}
          isValid={false}
          label="Suffix"
          maxLength={1}
          name="suffix"
          onBlur={[Function]}
          onChange={[Function]}
          validationMessage="Suffix cannot have special characters"
          value=""
        >
          <PPMSFormGroup
            error={false}
            id="formGroup-suffix"
          >
            <div
              className="usa-form-group"
              data-testid="formGroup"
              id="formGroup-suffix"
            >
              <PPMSLabel
                hint="(Optional)"
                htmlFor="suffix"
              >
                <label
                  className="usa-label"
                  data-testid="label"
                  htmlFor="suffix"
                >
                   
                  Suffix
                   
                   
                  <span
                    className="usa-hint"
                  >
                     
                    (Optional)
                     
                  </span>
                </label>
              </PPMSLabel>
              <PPMSTextInput
                autoComplete="off"
                disabled={false}
                id="suffix"
                inputRef={
                  Object {
                    "current": <input
                      autocomplete="off"
                      class="usa-input"
                      data-testid="textInput"
                      id="suffix"
                      maxlength="1"
                      name="suffix"
                      type="text"
                      value=""
                    />,
                  }
                }
                maxLength={1}
                name="suffix"
                onBlur={[Function]}
                onChange={[Function]}
                required={false}
                type="text"
                validationStatus=""
                value=""
              >
                <input
                  autoComplete="off"
                  className="usa-input"
                  data-testid="textInput"
                  disabled={false}
                  id="suffix"
                  key="key-suffix"
                  maxLength={1}
                  name="suffix"
                  onBlur={[Function]}
                  onChange={[Function]}
                  required={false}
                  type="text"
                  value=""
                />
              </PPMSTextInput>
            </div>
          </PPMSFormGroup>
        </PPMSInput>
      </div>
    </div>
    <div
      className="grid-row update-buttons"
    >
      <div
        className="tablet:grid-col-4"
      >
        <PPMSTooltip
          id="confirm-icn"
          placement="right"
          tooltipContent="ICN cannot be changed once confirmed."
          trigger="focus"
          triggerSource={
            <button
              className="usa-button"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Confirm ICN
            </button>
          }
        >
          <OverlayTrigger
            defaultShow={false}
            key="confirm-icn"
            overlay={
              <ForwardRef
                id="ppms-tooltip-confirm-icn"
                placement="right"
              >
                ICN cannot be changed once confirmed.
              </ForwardRef>
            }
            placement="right"
            trigger={
              Array [
                "hover",
                "focus",
              ]
            }
          >
            <RefHolder>
              <button
                className="usa-button"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                type="button"
              >
                Confirm ICN
              </button>
            </RefHolder>
            <Overlay
              flip={false}
              onHide={[Function]}
              placement="right"
              popperConfig={Object {}}
              rootClose={false}
              show={false}
              target={[Function]}
              transition={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": Object {
                    "appear": false,
                    "in": false,
                    "mountOnEnter": false,
                    "timeout": 300,
                    "unmountOnExit": false,
                  },
                  "displayName": "Fade",
                  "render": [Function],
                }
              }
            >
              <Overlay
                flip={false}
                onHide={[Function]}
                placement="right"
                popperConfig={
                  Object {
                    "modifiers": Array [
                      Object {
                        "name": "offset",
                        "options": Object {
                          "offset": [Function],
                        },
                      },
                      Object {
                        "effect": [Function],
                        "enabled": true,
                        "name": "popoverArrowMargins",
                        "phase": "main",
                        "requiresIfExists": Array [
                          "arrow",
                        ],
                      },
                    ],
                  }
                }
                rootClose={false}
                show={false}
                target={[Function]}
                transition={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "defaultProps": Object {
                      "appear": false,
                      "in": false,
                      "mountOnEnter": false,
                      "timeout": 300,
                      "unmountOnExit": false,
                    },
                    "displayName": "Fade",
                    "render": [Function],
                  }
                }
              />
            </Overlay>
          </OverlayTrigger>
        </PPMSTooltip>
      </div>
      <div
        className="tablet:grid-col-8"
      >
        <PPMSAlert
          alertBody=""
          alertClassName="email-verification-error"
          alertKey="email-verification-error"
          alertVariant="danger"
          id="icn-verification-danger-msg"
          show={false}
        >
          <Alert
            className="usa-alert usa-alert--error "
            closeLabel="Close alert"
            id="icn-verification-danger-msg"
            key="email-verification-error"
            show={false}
            transition={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "appear": false,
                  "in": false,
                  "mountOnEnter": false,
                  "timeout": 300,
                  "unmountOnExit": false,
                },
                "displayName": "Fade",
                "render": [Function],
              }
            }
          >
            <Fade
              appear={false}
              id="icn-verification-danger-msg"
              in={false}
              mountOnEnter={false}
              timeout={300}
              unmountOnExit={true}
            >
              <Transition
                addEndListener={[Function]}
                appear={false}
                enter={true}
                exit={true}
                id="icn-verification-danger-msg"
                in={false}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={true}
              />
            </Fade>
          </Alert>
        </PPMSAlert>
        <PPMSAlert
          alertBody=""
          alertClassName="email-verification-error"
          alertKey="email-verification-error"
          alertVariant="success"
          id="icn-verification-success-msg"
          show={false}
        >
          <Alert
            className="usa-alert usa-alert--success "
            closeLabel="Close alert"
            id="icn-verification-success-msg"
            key="email-verification-error"
            show={false}
            transition={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "appear": false,
                  "in": false,
                  "mountOnEnter": false,
                  "timeout": 300,
                  "unmountOnExit": false,
                },
                "displayName": "Fade",
                "render": [Function],
              }
            }
          >
            <Fade
              appear={false}
              id="icn-verification-success-msg"
              in={false}
              mountOnEnter={false}
              timeout={300}
              unmountOnExit={true}
            >
              <Transition
                addEndListener={[Function]}
                appear={false}
                enter={true}
                exit={true}
                id="icn-verification-success-msg"
                in={false}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={true}
              />
            </Fade>
          </Alert>
        </PPMSAlert>
      </div>
    </div>
  </IcnClass>
</PropertyProvider>
`;
